### PR TITLE
feat(button): add label text wrap option

### PR DESCRIPTION
**Pull request by aramos-adobe**
## Description

- This change makes buttons _not_ wrap label text by default.
- Introducing a class to make label text wrap if dev chooses to do so.
- Added label wrap class in Text Overflow story in Storybook

## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Regression testing

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [ ] The pages render correctly, are accessible, and are responsive.

2. If components have been modified, VRTs have been run on this branch:

- [ ] VRTs have been run and looked at.
- [ ] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.

## To-do list

- [ ] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [ ] I have updated relevant storybook stories and templates.
- [ ] I have tested these changes in Windows High Contrast mode.
- [ ] If my change impacts **other components**, I have tested to make sure they don't break.
- [ ] If my change impacts **documentation**, I have updated the documentation accordingly.
- [ ] ✨ This pull request is ready to merge. ✨

### CODE REVIEW

**Comment by aramos-adobe** [comment]
Screenshot 2024-09-11 at 9 27 06 AM
Screenshot 2024-09-11 at 9 27 15 AM
Screenshot 2024-09-11 at 9 27 29 AM

File: .changeset/mean-shrimps-press.md
Comment (jawinn) [inline .changeset/mean-shrimps-press.md:1]: I see two changesets here with similar content; it looks like we only need one?

File: components/button/stories/template.js [hunk 1]
```diff
@@ -158,13 +160,15 @@ export const TextOverflowTemplate = (args) => Container({
       customStyles: {
         "max-inline-size": "480px",
       },
+      isLabelWrapped: true,
```
Comment (jawinn) [inline components/button/stories/template.js:163]:
```suggestion
```
It looks like this was an older variable naming?

File: components/button/stories/template.js [hunk 2]
```diff
@@ -158,13 +160,15 @@ export const TextOverflowTemplate = (args) => Container({
...
       label: "An example of text overflow behavior when there is no icon. When the button text is too long for the horizontal space available, it wraps to form another line.",
     })}
     ${Template({
       ...args,
       customStyles: {
         "max-inline-size": "480px",
       },
+      isLabelWrapped: true,
```
Comment (jawinn) [inline components/button/stories/template.js:171]:
```suggestion
```

File: components/button/stories/button.stories.js [hunk 3]
```diff
@@ -58,6 +58,13 @@ export default {
 		isActive,
 		isPending,
 		staticColor,
+		noWrap: {
+			name: "Disable Label Wrap",
+			type: { name: "boolean" },
+			table: {
+				category: "Content",
```
Comment (jawinn) [inline components/button/stories/button.stories.js:65]:
```suggestion
				category: "Advanced",
```
I would categorize this as advanced, since it's not a regular option from the spec and isn't typically recommended. Otherwise it would be in the "Component" category. "Content" is typically for controls that are for modifying the content displayed within the story (e.g. text, image source, icon name, etc).

**Review by jawinn — CHANGES_REQUESTED** [review]
I think the `nowrap` CSS should work here. This and the added control look good!

Could you make sure that we have this included in the VRTs, by adding a `noWrap: true` test in the `button.test.js` file? It likely will need a set width on the wrapperStyles, or a min-inline-size like the text overflow examples in order for wrapping to happen. When added it'll be viewable on the Default story with "Show testing preview" turned on.

File: components/button/stories/button.stories.js [hunk 4]
```diff
@@ -58,6 +58,13 @@ export default {
 		isActive,
 		isPending,
 		staticColor,
+		noWrap: {
+			name: "Disable Label Wrap",
```
Comment (jawinn) [inline components/button/stories/button.stories.js:62]:
```suggestion
			name: "Disable label wrap",
```
Use [sentence case](https://spectrum.adobe.com/page/grammar-and-mechanics/#Sentence-case) on this. That's something we've been working on making more consistent across the docs lately.

File: components/button/stories/template.js (hunk 1 above)
Comment (aramos-adobe) [reply components/button/stories/template.js:163]: Yes I need to get rid of this. This was from a previous approach of making the button label wrap by default.

File: components/button/index.css [hunk 5]
```diff
@@ -294,6 +294,17 @@ a.spectrum-Button {
 	}
 }
 
+/* Disable button label wrap  */
+.spectrum-Button--noWrap {
+	justify-content: flex-start;
```
Comment (jawinn) [inline components/button/index.css:299]: With the `overflow` and `text-overflow` additions, is this addition still needed? With short text plus the minimum width and for buttons with some other set width, the text will no longer be centered:
Screenshot 2024-09-16 at 3 10 40 PM

File: .changeset/green-maps-end.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"@spectrum-css/button": minor
+---
+
+Adding the ability to disable label wrap in button component
```
Comment (jawinn) [inline .changeset/green-maps-end.md:5]:
```suggestion
Adds the ability to disable text wrapping within the button component's label by applying the class `.spectrum-Button--noWrap` to `.spectrum-Button`.
```
Changing the tense and adding some more detail here, since this will go in the change log and be referenced by SWC and others.

File: components/button/index.css [hunk 7]
```diff
@@ -294,6 +294,17 @@ a.spectrum-Button {
...
+	
+	.spectrum-Button-label {
+		white-space: nowrap;
+		overflow: hidden;
+		text-overflow: ellipsis;
```
Comment (jawinn) [inline components/button/index.css:304]: Now that this is working like the text overflow behavior of Action button, it creates accessibility issues when there is text overflow, since the text can be sometimes unreadable. Does that mean we need to provide similar guidelines about a tooltip? We'd be asking more of the implementation for this outside the spec feature: https://spectrum.adobe.com/page/action-button/#Text-overflow

File: components/button/stories/button.test.js [hunk 8]
```diff
@@ -66,6 +66,27 @@ export const ButtonGroups = Variants({
...
+			testHeading: "Disable label wrapping with workflow icon",
+			customStyles: {
+				"max-inline-size": "120px",
+			},
+			iconName: "Star",
+			label: "Be a premium member",
+			withStates: false,
+			noWrap: true,
+			Template,
+		},
```
Comment (jawinn) [inline components/button/stories/button.test.js:89]: Thanks for making sure this is represented in the VRTs 👍

**Review by jawinn — CHANGES_REQUESTED** [review]

File: components/button/index.css (hunk 5 above)
Comment (aramos-adobe) [reply components/button/index.css:299]: @jawinn When the max-inline-size is set all the button content gets cut off from left and right edges of the container. I was thinking of doing a flex-start and overflow additions it'll show the dev the text is too long and they need to make an adjustment character wise. 

After some thought, I'm going to remove this because it's not a design specification and I don't want teams to get mislead in anyway.

Comment (jawinn) [reply components/button/index.css]: I do think what you had, `.spectrum-Button-label { overflow: hidden; text-overflow: ellipsis; }` is a good idea. I think we _should_ handle the text overflow for cases where the button's width is constrained, as it gets a bit broken as seen in the Chromatic testing view right now. Does it work okay with just those two properties?

Comment (aramos-adobe) [reply components/button/index.css]: @jawinn Yes, they work with just these two properties.

File: components/button/index.css (hunk 7 above)
Comment (aramos-adobe) [reply components/button/index.css:301]: @jawinn I tried it and it added the test case. 

Tooltip only shows when `nowrap` is true and `hideLabel` is false. 
If the text is not "ellipsized" or visible I added a class to hide the tooltip: `spectrum-Tooltip--hide`.

File: components/button/stories/button.stories.js [hunk 9]
```diff
@@ -286,3 +295,18 @@ WithWrapping.args = {
 WithWrapping.parameters = {
 	chromatic: { disableSnapshot: true },
 };
+
+/**
+ * The normal behavior for lengthy text in the given horizontal space available is that it will wrap to form another line. By using the `noWrap` option, the lengthy button text will horizontally expand the entirety of the container. Be sure to define the width of the button container using this option. Please note: this can cause undesired overflow experiences and to help prevent this, the overflowing text will attempt to hide by showing an ellipsis (...).
```
Comment (jawinn) [inline components/button/stories/button.stories.js:300]:
```suggestion
The normal behavior for lengthy text in the given horizontal space available is that it will wrap to form another line. By using the `.spectrum-Button--noWrap` class, the lengthy button text will not cause a line break and the width of the button will expand until it reaches its maximum width.

Please note: this can cause undesired overflow experiences and to help prevent this, the overflowing text will attempt to hide by showing an ellipsis (...). This is demonstrated in the last two examples below, by constraining the maximum width of the button.

This option is not part of the design spec, so please use carefully, with consideration of the overflow behavior and the readability of the button's content.
```

File: components/button/stories/button.stories.js [hunk 10]
```diff
@@ -286,3 +295,18 @@ WithWrapping.args = {
 WithWrapping.parameters = {
 	chromatic: { disableSnapshot: true },
 };
+
+/**
+ * The normal behavior for lengthy text in the given horizontal space available is that it will wrap to form another line. By using the `noWrap` option, the lengthy button text will horizontally expand the entirety of the container. Be sure to define the width of the button container using this option. Please note: this can cause undesired overflow experiences and to help prevent this, the overflowing text will attempt to hide by showing an ellipsis (...). 
+ * */
+
+export const DisableWrapping = TextWrapTemplate.bind({});
```
Comment (jawinn) [inline components/button/stories/button.stories.js:303]: What do you think about showing a third option which demonstrates that this can escape its container? I experimented in the inspector a little bit:
Screenshot 2024-09-19 at 3 10 34 PM
Here I'm setting the width on the inner container (adjustable through `wrapperStyles`), and the buttons with ellipsis are `max-inline-size: 100%`.

This makes me wonder if we need a new mod for `max-inline-size`, if we're expecting users in SWC to be able to want to change that property too (@pfulton)?

**Review by jawinn — CHANGES_REQUESTED** [review]

**Review by pfulton — APPROVED** [review]
Thanks!!

**Review by jawinn — APPROVED** [review]

**Comment by aramos-adobe** [comment]
<iframe src="https://giphy.com/embed/gKfyusl0PRPdTNmwnD" width="480" height="250" style="" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/pleading-vincent-van-gogh-please-god-gKfyusl0PRPdTNmwnD">via GIPHY</a></p>

**Comment by aramos-adobe** [comment]
image
